Source/WebCore/html/LinkRelAttribute.cpp

```diff
@@ -40,67 +40,85 @@
-    else if (equalLettersIgnoringASCIICase(rel, "dns-prefetch"_s))
-        isDNSPrefetch = true;
-    else if (document.settings().linkPreconnectEnabled() && equalLettersIgnoringASCIICase(rel, "preconnect"_s))
-        isLinkPreconnect = true;
-    else if (document.settings().linkModulePreloadEnabled() && equalLettersIgnoringASCIICase(rel, "modulepreload"_s))
-        isLinkModulePreload = true;
-    else if (document.settings().linkPreloadEnabled() && equalLettersIgnoringASCIICase(rel, "preload"_s))
-        isLinkPreload = true;
-    else if (document.settings().linkPrefetchEnabled() && equalLettersIgnoringASCIICase(rel, "prefetch"_s))
-        isLinkPrefetch = true;
     else if (equalLettersIgnoringASCIICase(rel, "alternate stylesheet"_s) || equalLettersIgnoringASCIICase(rel, "stylesheet alternate"_s)) {
         isStyleSheet = true;
         isAlternate = true;
-#if ENABLE(APPLICATION_MANIFEST)
-    } else if (equalLettersIgnoringASCIICase(rel, "manifest"_s)) {
-        isApplicationManifest = true;
-#endif
-    } else if (equalLettersIgnoringASCIICase(rel, "expect"_s))
-        isInternalResourceLink = true;
-    else {
+    } else {
         // Tokenize the rel attribute and set bits based on specific keywords that we find.
-        for (auto line : rel.split('\n')) {
-            for (auto word : line.split(' ')) {
-                if (equalLettersIgnoringASCIICase(word, "stylesheet"_s))
-                    isStyleSheet = true;
-                else if (equalLettersIgnoringASCIICase(word, "alternate"_s))
-                    isAlternate = true;
-                else if (equalLettersIgnoringASCIICase(word, "icon"_s))
-                    iconType = LinkIconType::Favicon;
-                else if (equalLettersIgnoringASCIICase(word, "apple-touch-icon"_s))
-                    iconType = LinkIconType::TouchIcon;
-                else if (equalLettersIgnoringASCIICase(word, "apple-touch-icon-precomposed"_s))
-                    iconType = LinkIconType::TouchPrecomposedIcon;
+        unsigned length = rel.length();
+        unsigned start = 0;
+        while (start < length) {
+            if (isASCIIWhitespace(rel[start])) {
+                start++;
+                continue;
+            }
+            unsigned end = start + 1;
+            while (end < length && !isASCIIWhitespace(rel[end]))
+                end++;
+
+            auto possibleLinkType = rel.substring(start, end - start);
+            start = end;
+
+            auto found = false;
+            for (auto linkType : linkRelAttributeMap.keys()) {
+                if (equalLettersIgnoringASCIICase(possibleLinkType, linkType)) {
+                    linkRelAttributeMap.get(linkType)(*this);
+                    found = true;
+                }
             }
+            if (found)
+                continue;
+
+            if (document.settings().linkPreconnectEnabled() && equalLettersIgnoringASCIICase(possibleLinkType, "preconnect"_s))
```

Comment:
Since there are no chance of these entries ever matching other entires in the map,
we might as well as use the same lookup table for all these runtime enabled features,
and check the runtime enabled-ness optionally:

```cpp
auto found = ([&] {
    for (auto& entry : linkTypeTable) {
        if (equalLettersIgnoringASCIICase(entry.name, possibleLinkType) && entry.isEnabled()) {
            entry.updateRelAttribute(*this);
            return true;
        }
    }
    return false;
})();
```